File: src/other-languages/java/tests/LinkProvenanceTest.java
// Copyright 2026 Daniel Smith
// Licensed under the Apache License, Version 2.0
// See https://www.apache.org/licenses/LICENSE-2.0

package tests;

import alap.LinkProvenance;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Tests for LinkProvenance — tier stamping + predicates + cloneTo.
 *
 * <p>Run: {@code java -ea tests.LinkProvenanceTest}
 */
public class LinkProvenanceTest {

    private static int passed = 0;
    private static int failed = 0;

    public static void main(String[] args) {
        System.out.println("Alap Java — LinkProvenance Test Suite\n");

        section("Stamp + get");
        run(LinkProvenanceTest::testStampAuthorThenRead);
        run(LinkProvenanceTest::testStampStorageLocal);
        run(LinkProvenanceTest::testStampStorageRemote);
        run(LinkProvenanceTest::testStampProtocol);
        run(LinkProvenanceTest::testUnstampedReturnsNull);
        run(LinkProvenanceTest::testStampOverwritesExisting);
        run(LinkProvenanceTest::testStampUsesReservedKey);

        section("Invalid tier rejection");
        run(LinkProvenanceTest::testRejectsUnknownTier);
        run(LinkProvenanceTest::testRejectsTypoAuthor);
        run(LinkProvenanceTest::testRejectsEmpty);
        run(LinkProvenanceTest::testRejectsNull);
        run(LinkProvenanceTest::testRejectsBareProtocolPrefix);
        run(LinkProvenanceTest::testAcceptsAnyProtocolSuffix);

        section("Tier predicates");
        run(LinkProvenanceTest::testAuthorTruePredicate);
        run(LinkProvenanceTest::testStorageTrueForLocal);
        run(LinkProvenanceTest::testStorageTrueForRemote);
        run(LinkProvenanceTest::testProtocolTrueForProtocolWeb);
        run(LinkProvenanceTest::testAllFalseForUnstamped);

        section("cloneTo");
        run(LinkProvenanceTest::testCloneToCopiesStamp);
        run(LinkProvenanceTest::testCloneToNoOpWhenSrcUnstamped);
        run(LinkProvenanceTest::testCloneToOverwritesExistingDest);

        System.out.println("\n==================================================");
        System.out.println(passed + " passed, " + failed + " failed, " + (passed + failed) + " total");
        System.exit(failed == 0 ? 0 : 1);
    }

    private static Map<String, Object> link() {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("url", "/a");
        return m;
    }

    // --- Stamp + get ---

    static void testStampAuthorThenRead() {
        Map<String, Object> m = link();
        LinkProvenance.stamp(m, "author");
        assertEquals("author", LinkProvenance.get(m), "author stamp read");
    }

    static void testStampStorageLocal() {
        Map<String, Object> m = link();
        LinkProvenance.stamp(m, "storage:local");
        assertEquals("storage:local", LinkProvenance.get(m), "storage:local stamp read");
    }

    static void testStampStorageRemote() {
        Map<String, Object> m = link();
        LinkProvenance.stamp(m, "storage:remote");
        assertEquals("storage:remote", LinkProvenance.get(m), "storage:remote stamp read");
    }

    static void testStampProtocol() {
        Map<String, Object> m = link();
        LinkProvenance.stamp(m, "protocol:web");
        assertEquals("protocol:web", LinkProvenance.get(m), "protocol:web stamp read");
    }

    static void testUnstampedReturnsNull() {
        assertEquals(null, LinkProvenance.get(link()), "unstamped returns null");
    }

    static void testStampOverwritesExisting() {
        Map<String, Object> m = link();
        LinkProvenance.stamp(m, "author");
        LinkProvenance.stamp(m, "protocol:web");
        assertEquals("protocol:web", LinkProvenance.get(m), "second stamp overwrites");
    }

    static void testStampUsesReservedKey() {
        Map<String, Object> m = link();
        LinkProvenance.stamp(m, "author");
        assertTrue(m.containsKey(LinkProvenance.PROVENANCE_KEY), "reserved key present");
        assertEquals("author", m.get(LinkProvenance.PROVENANCE_KEY), "reserved key value");
    }

    // --- Invalid tier ---

    static void testRejectsUnknownTier() {
        assertThrows(IllegalArgumentException.class,
            () -> LinkProvenance.stamp(link(), "admin"), "rejects admin");
    }

    static void testRejectsTypoAuthor() {
        assertThrows(IllegalArgumentException.class,
            () -> LinkProvenance.stamp(link(), "Author"), "rejects Author");
    }

    static void testRejectsEmpty() {
        assertThrows(IllegalArgumentException.class,
            () -> LinkProvenance.stamp(link(), ""), "rejects empty");
    }

    static void testRejectsNull() {
        assertThrows(IllegalArgumentException.class,
            () -> LinkProvenance.stamp(link(), null), "rejects null");
    }

    static void testRejectsBareProtocolPrefix() {
        assertThrows(IllegalArgumentException.class,
            () -> LinkProvenance.stamp(link(), "protocol:"), "rejects bare protocol:");
    }

    static void testAcceptsAnyProtocolSuffix() {
        Map<String, Object> m = link();
        LinkProvenance.stamp(m, "protocol:custom_handler_42");
        assertEquals("protocol:custom_handler_42", LinkProvenance.get(m), "accepts custom suffix");
    }

    // --- Predicates ---

    static void testAuthorTruePredicate() {
        Map<String, Object> m = link();
        LinkProvenance.stamp(m, "author");
        assertTrue(LinkProvenance.isAuthorTier(m), "isAuthorTier true");
        assertFalse(LinkProvenance.isStorageTier(m), "isStorageTier false");
        assertFalse(LinkProvenance.isProtocolTier(m), "isProtocolTier false");
    }

    static void testStorageTrueForLocal() {
        Map<String, Object> m = link();
        LinkProvenance.stamp(m, "storage:local");
        assertFalse(LinkProvenance.isAuthorTier(m), "isAuthorTier false");
        assertTrue(LinkProvenance.isStorageTier(m), "isStorageTier true");
        assertFalse(LinkProvenance.isProtocolTier(m), "isProtocolTier false");
    }

    static void testStorageTrueForRemote() {
        Map<String, Object> m = link();
        LinkProvenance.stamp(m, "storage:remote");
        assertTrue(LinkProvenance.isStorageTier(m), "storage:remote -> isStorageTier true");
    }

    static void testProtocolTrueForProtocolWeb() {
        Map<String, Object> m = link();
        LinkProvenance.stamp(m, "protocol:web");
        assertFalse(LinkProvenance.isAuthorTier(m), "isAuthorTier false");
        assertFalse(LinkProvenance.isStorageTier(m), "isStorageTier false");
        assertTrue(LinkProvenance.isProtocolTier(m), "isProtocolTier true");
    }

    static void testAllFalseForUnstamped() {
        Map<String, Object> m = link();
        assertFalse(LinkProvenance.isAuthorTier(m), "unstamped !author");
        assertFalse(LinkProvenance.isStorageTier(m), "unstamped !storage");
        assertFalse(LinkProvenance.isProtocolTier(m), "unstamped !protocol");
    }

    // --- cloneTo ---

    static void testCloneToCopiesStamp() {
        Map<String, Object> src = link();
        LinkProvenance.stamp(src, "protocol:web");
        Map<String, Object> dest = new LinkedHashMap<>();
        dest.put("url", "/b");
        LinkProvenance.cloneTo(src, dest);
        assertEquals("protocol:web", LinkProvenance.get(dest), "cloneTo copies");
    }

    static void testCloneToNoOpWhenSrcUnstamped() {
        Map<String, Object> src = link();
        Map<String, Object> dest = new HashMap<>();
        dest.put("url", "/b");
        LinkProvenance.cloneTo(src, dest);
        assertEquals(null, LinkProvenance.get(dest), "dest still unstamped");
        assertFalse(dest.containsKey(LinkProvenance.PROVENANCE_KEY), "reserved key not set");
    }

    static void testCloneToOverwritesExistingDest() {
        Map<String, Object> src = link();
        LinkProvenance.stamp(src, "storage:remote");
        Map<String, Object> dest = new LinkedHashMap<>();
        dest.put("url", "/b");
        LinkProvenance.stamp(dest, "author");
        LinkProvenance.cloneTo(src, dest);
        assertEquals("storage:remote", LinkProvenance.get(dest), "cloneTo overwrites");
    }

    // --- Runner ---

    private static void run(Runnable test) {
        try {
            test.run();
            passed++;
        } catch (AssertionError e) {
            failed++;
            System.out.println("  FAIL: " + e.getMessage());
        } catch (Throwable t) {
            failed++;
            System.out.println("  ERROR: " + t.getClass().getSimpleName() + ": " + t.getMessage());
        }
    }

    private static void section(String name) {
        System.out.println("\n--- " + name + " ---");
    }

    private static void assertEquals(Object expected, Object actual, String msg) {
        if (expected == null ? actual != null : !expected.equals(actual)) {
            throw new AssertionError(msg + ": expected " + expected + " but got " + actual);
        }
        System.out.println("  PASS: " + msg);
    }

    private static void assertTrue(boolean cond, String msg) {
        if (!cond) throw new AssertionError(msg + ": expected true");
        System.out.println("  PASS: " + msg);
    }

    private static void assertFalse(boolean cond, String msg) {
        if (cond) throw new AssertionError(msg + ": expected false");
        System.out.println("  PASS: " + msg);
    }

    private static void assertThrows(Class<? extends Throwable> expected, Runnable action, String msg) {
        try {
            action.run();
        } catch (Throwable t) {
            if (expected.isInstance(t)) {
                System.out.println("  PASS: " + msg);
                return;
            }
            throw new AssertionError(msg + ": expected " + expected.getSimpleName() +
                " but got " + t.getClass().getSimpleName());
        }
        throw new AssertionError(msg + ": expected " + expected.getSimpleName() + " but nothing was thrown");
    }
}

File: src/other-languages/ruby/test/link_provenance_test.rb
# frozen_string_literal: true

# Copyright 2026 Daniel Smith
# Licensed under the Apache License, Version 2.0
# See https://www.apache.org/licenses/LICENSE-2.0

require "minitest/autorun"
require_relative "../lib/alap"

class LinkProvenanceStampAndGetTest < Minitest::Test
  def test_stamp_author_then_read
    link = { "url" => "/a" }
    Alap::LinkProvenance.stamp(link, "author")
    assert_equal "author", Alap::LinkProvenance.get(link)
  end

  def test_stamp_storage_local
    link = { "url" => "/a" }
    Alap::LinkProvenance.stamp(link, "storage:local")
    assert_equal "storage:local", Alap::LinkProvenance.get(link)
  end

  def test_stamp_storage_remote
    link = { "url" => "/a" }
    Alap::LinkProvenance.stamp(link, "storage:remote")
    assert_equal "storage:remote", Alap::LinkProvenance.get(link)
  end

  def test_stamp_protocol
    link = { "url" => "/a" }
    Alap::LinkProvenance.stamp(link, "protocol:web")
    assert_equal "protocol:web", Alap::LinkProvenance.get(link)
  end

  def test_unstamped_returns_nil
    assert_nil Alap::LinkProvenance.get({ "url" => "/a" })
  end

  def test_stamp_overwrites_existing
    link = { "url" => "/a" }
    Alap::LinkProvenance.stamp(link, "author")
    Alap::LinkProvenance.stamp(link, "protocol:web")
    assert_equal "protocol:web", Alap::LinkProvenance.get(link)
  end

  def test_stamp_uses_reserved_key
    link = { "url" => "/a" }
    Alap::LinkProvenance.stamp(link, "author")
    assert link.key?(Alap::LinkProvenance::PROVENANCE_KEY)
    assert_equal "author", link[Alap::LinkProvenance::PROVENANCE_KEY]
  end
end

class LinkProvenanceInvalidTierTest < Minitest::Test
  def test_rejects_unknown_tier
    assert_raises(ArgumentError) { Alap::LinkProvenance.stamp({ "url" => "/a" }, "admin") }
  end

  def test_rejects_typo_author
    assert_raises(ArgumentError) { Alap::LinkProvenance.stamp({ "url" => "/a" }, "Author") }
  end

  def test_rejects_empty
    assert_raises(ArgumentError) { Alap::LinkProvenance.stamp({ "url" => "/a" }, "") }
  end

  def test_rejects_non_string
    assert_raises(ArgumentError) { Alap::LinkProvenance.stamp({ "url" => "/a" }, 42) }
  end

  def test_accepts_any_protocol_suffix
    link = { "url" => "/a" }
    Alap::LinkProvenance.stamp(link, "protocol:custom_handler_42")
    assert_equal "protocol:custom_handler_42", Alap::LinkProvenance.get(link)
  end
end

class LinkProvenanceTierPredicatesTest < Minitest::Test
  def test_author_true_for_author
    link = { "url" => "/a" }
    Alap::LinkProvenance.stamp(link, "author")
    assert Alap::LinkProvenance.author_tier?(link)
    refute Alap::LinkProvenance.storage_tier?(link)
    refute Alap::LinkProvenance.protocol_tier?(link)
  end

  def test_storage_true_for_local
    link = { "url" => "/a" }
    Alap::LinkProvenance.stamp(link, "storage:local")
    refute Alap::LinkProvenance.author_tier?(link)
    assert Alap::LinkProvenance.storage_tier?(link)
    refute Alap::LinkProvenance.protocol_tier?(link)
  end

  def test_storage_true_for_remote
    link = { "url" => "/a" }
    Alap::LinkProvenance.stamp(link, "storage:remote")
    assert Alap::LinkProvenance.storage_tier?(link)
  end

  def test_protocol_true_for_protocol_web
    link = { "url" => "/a" }
    Alap::LinkProvenance.stamp(link, "protocol:web")
    refute Alap::LinkProvenance.author_tier?(link)
    refute Alap::LinkProvenance.storage_tier?(link)
    assert Alap::LinkProvenance.protocol_tier?(link)
  end

  def test_all_false_for_unstamped
    link = { "url" => "/a" }
    refute Alap::LinkProvenance.author_tier?(link)
    refute Alap::LinkProvenance.storage_tier?(link)
    refute Alap::LinkProvenance.protocol_tier?(link)
  end
end

class LinkProvenanceCloneToTest < Minitest::Test
  def test_copies_stamp
    src = { "url" => "/a" }
    Alap::LinkProvenance.stamp(src, "protocol:web")
    dest = { "url" => "/b" }
    Alap::LinkProvenance.clone_to(src, dest)
    assert_equal "protocol:web", Alap::LinkProvenance.get(dest)
  end

  def test_no_op_when_src_unstamped
    src = { "url" => "/a" }
    dest = { "url" => "/b" }
    Alap::LinkProvenance.clone_to(src, dest)
    assert_nil Alap::LinkProvenance.get(dest)
    refute dest.key?(Alap::LinkProvenance::PROVENANCE_KEY)
  end

  def test_overwrites_existing_stamp_on_dest
    src = { "url" => "/a" }
    Alap::LinkProvenance.stamp(src, "storage:remote")
    dest = { "url" => "/b" }
    Alap::LinkProvenance.stamp(dest, "author")
    Alap::LinkProvenance.clone_to(src, dest)
    assert_equal "storage:remote", Alap::LinkProvenance.get(dest)
  end
end
